builder/bundles.go

```diff
@@ -480,30 +471,146 @@ func downloadRpms(packagerCmd, rpmList []string, baseDir string, maxRetries int)
 	return nil, downloadErr
 }
 
-func installBundleToFull(packagerCmd []string, buildVersionDir string, bundle *bundle, downloadRetries int) error {
+func downloadRpm(packagerCmd []string, rpmList []string, baseDir string, downloadRetries int) (error, string) {
+	// Retry RPM downloads to avoid timeout failures due to slow network
+	_, err := downloadRpms(packagerCmd, rpmList, baseDir, downloadRetries)
+	if err != nil {
+		return err, ""
+	}
+	path := baseDir + rpmDir
+
+	fullRpmPath, err := filepath.Glob(filepath.Join(path, "clear-*"))
+	if err != nil {
+		fmt.Print("cant find path to rpm")
+		return err, ""
+	}
+	if len(fullRpmPath) < 0 {
+		return fmt.Errorf("cant find rpms packages"), ""
+	}
+	fullRpmPath[0] = fullRpmPath[0] + "/packages"
+
+	return nil, fullRpmPath[0]
+}
+
+func extractRpm(baseDir string, rpm string) error {
+	cmd := exec.Command("rpm2cpio", rpm)
+	cmd.Env = os.Environ()
+	cmd2 := exec.Command("cpio", "-idm")
+	cmd2.Env = os.Environ()
+	cmd2.Dir = baseDir
+
+	pr, pw := io.Pipe()
+	cmd.Stdout = pw
+	cmd2.Stdin = pr
+
 	var err error
-	baseDir := filepath.Join(buildVersionDir, "full")
-	rpmList := []string{}
-
-	if len(bundle.AllPackages) > 0 {
-		// There were packages directly included for this bundle so
-		// install to full chroot. This check is necessary so we don't
-		// call dnf install with no package listed.
-		for p := range bundle.AllPackages {
-			rpmList = append(rpmList, p)
+	err = cmd.Start()
+	if err != nil {
+		return err
+	}
+	err = cmd2.Start()
+	if err != nil {
+		return err
+	}
+	err = cmd.Wait()
+	if err != nil {
+		return err
+	}
+	err = pw.Close()
+	if err != nil {
+		return err
+	}
+	err = cmd2.Wait()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func installBundleToFull(packagerCmd []string, baseDir string, bundle *bundle, downloadRetries int, numWorkers int, localPath string) error {
+	var err error
+
+	var wg sync.WaitGroup
+	rpmCh := make(chan string)
+	errorCh := make(chan error, numWorkers)
+
+	if len(bundle.AllPackages) < numWorkers {
+		numWorkers = len(bundle.AllPackages)
+	}
+
+	wg.Add(numWorkers)
+
+	rpmWorker := func() {
+		for rpm := range rpmCh {
+			e := extractRpm(baseDir, rpm)
+			if e != nil {
+				errorCh <- e
+				break
+			}
 		}
+		wg.Done()
+	}
 
-		// Retry RPM downloads to avoid timeout failures due to slow network
-		_, err = downloadRpms(packagerCmd, rpmList, baseDir, downloadRetries)
-		if err != nil {
-			return err
+	for i := 0; i < numWorkers; i++ {
+		go rpmWorker()
+	}
+
+	// feed the channel
+	var errCh error
+	var rpmList []string
+	for rpm, present := range bundle.AllPackages {
+		if present {
+			if rpmMap[rpm] {
```

Comment:
If rpm is the map key, will there ever be a case where you will hit this condition?